File: 0725-split-linked-list-in-parts/0725-split-linked-list-in-parts.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def splitListToParts(self, head, k):
        length = 0
        curr = head
        while curr:
            length += 1
            curr = curr.next

        part_size = length // k
        extra_parts = length % k

        result = []
        curr = head
        for i in range(k):
            size = part_size + (1 if extra_parts > 0 else 0)
            if size > 0:
                result.append(curr)
            else:
                result.append(None)
            
            for j in range(size - 1):
                if curr:
                    curr = curr.next

            if curr:
                next_part = curr.next
                curr.next = None
                curr = next_part
 
            if extra_parts > 0:
                extra_parts -= 1

        return result
        

File: 0725-split-linked-list-in-parts/0725-split-linked-list-in-parts.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} k
 * @return {ListNode[]}
 */
var splitListToParts = function(head, k) {
    let length = 0;
    let curr = head;
    
    while (curr !== null) {
        length++;
        curr = curr.next;
    }
 
    const partSize = Math.floor(length / k);
    let extraParts = length % k;
    
    const result = [];
    curr = head;
    
    for (let i = 0; i < k; i++) {
        let size = partSize + (extraParts > 0 ? 1 : 0);
        if (size > 0) {
            result.push(curr);
        } else {
            result.push(null);
        }

        for (let j = 1; j < size; j++) {
            if (curr !== null) {
                curr = curr.next;
            }
        }
        
        if (curr !== null) {
            let nextPart = curr.next;
            curr.next = null;
            curr = nextPart;
        }
        
        if (extraParts > 0) {
            extraParts--;
        }
    }
    
    return result;
};